daemon/cmd/status.go

```diff
@@ -388,9 +388,19 @@ func (c *clusterNodesClient) NodeAdd(newNode nodeTypes.Node) error {
 
 func (c *clusterNodesClient) NodeUpdate(oldNode, newNode nodeTypes.Node) error {
 	c.Lock()
+	defer c.Unlock()
+
+	// If the node is on the added list, just update it
+	for i, added := range c.NodesAdded {
+		if added.Name == newNode.Fullname() {
+			c.NodesAdded[i] = newNode.GetModel()
+			return nil
+		}
+	}
+
+	// otherwise, add the new node and remove the old one
```

Comment:
> Note that the setNodes() function on pkg/health/server/prober.go first
deletes the removed nodes and then adds the new ones, which means that
the old version of the node would be added and remain as stale on the
health server.

Can you elaborate on the timeline a bit more here? Do you mean that you think we are missing node delete events for some portion of the cluster?

Reply:
Just re-read through a few times and I get it now :-)

* Add (node-v1) -> v1 in c.Added
* Update (node-v2) -> v1 is still in c.Added because of the first step; we now add v1 to c.Removed and *also* add v2 to c.Added. So even though the health server removes the old version, it then readds it because the update didn't handle the update correctly.

Reply:
As far as I understand, `c.NodesAdded` and `c.NodesDeleted` maintain the nodes added and deleted from the last request of the client. Whenever the client requests the changes, the two structures above are cleared and are subsequently updated with new changes.

Let's say that the following things happen before client has a chnce to make a request:
 - Node A is added (A0)
 - Node A is updated (A1)

With the old code this will result in:

- NodeAdd(A0) -> NodesAdded: [A0], NodesDeleted: []
- NodeUpdate(A0, A1) -> NodesAdded: [A0, A1], NodesDeleted: [A0]

The code https://github.com/cilium/cilium/blob/326487aa506df7b384603cd403b2b2504fec4312/pkg/health/server/prober.go#L185 that adds them to the health server roughly does:

```
for node in NodesDeleted:
     - remove node
for node in NodesADded:
     - add node
```

Which means that A0 from the example above will be added to the nodes.